cmake/DefaultTargetProperties.cmake

```diff
@@ -0,0 +1,129 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2013-2022 The Foundry Visionmongers Ltd
+
+function(set_default_target_properties target_name)
+    #-------------------------------------------------------------------
+    # C++ standard
+
+    # Minimum C++ standard
+    target_compile_features(${target_name} PRIVATE cxx_std_14)
+
+    set_target_properties(
+        ${target_name}
+        PROPERTIES
+        # Ensure the proposed compiler supports our minimum C++
+        # standard.
+        CXX_STANDARD_REQUIRED ON
+        # Disable compiler extensions. E.g. use -std=c++11  instead of
+        # -std=gnu++11.  Helps limit cross-platform issues.
+        CXX_EXTENSIONS OFF
+    )
+
+    #-------------------------------------------------------------------
+    # Compiler warnings
+
+    include(CompilerWarnings)
+    set_default_compiler_warnings(${target_name})
+
+
+    #-------------------------------------------------------------------
+    # Potentially enable sanitizers.
+    include(Sanitizers)
+    enable_sanitizers(project_options)
+
+
+    #-------------------------------------------------------------------
+    # Symbol visibility
+
+    # Hide symbols from this library by default.
+    set_target_properties(
+        ${target_name}
+        PROPERTIES
+        C_VISIBILITY_PRESET hidden
+        CXX_VISIBILITY_PRESET hidden
+        VISIBILITY_INLINES_HIDDEN YES
+    )
+
+    # Hide all symbols from external statically linked libraries.
+    if (CMAKE_CXX_COMPILER_ID MATCHES ".*Clang" OR CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+        target_link_options(${target_name} PRIVATE -Wl,--exclude-libs,ALL)
+    endif ()
```

Comment:
I can't see any target_link_libraries that necessitate this.

Reply:
Not yet at least. The idea was to, by default, suppress exporting any symbols we don't explicitly want to.

I've been bitten by this before when two dsos both statically linked the same library (grpc), resulting in conflicting symbols and double-deletion crashes.

I figured it was better to start in "safe mode" and open up symbols if necessary later.

Reply:
Don't disagree. I would've just deferred adding it until it's needed.

I know you can't guarantee dependencies have been built correctly (unlike we force that in our Conan recipes), so I wasn't calling it out as bad - just unnecessary for this PR.